### ISSUE TITLE
openLicense must return an error code relative to the error type

### ISSUE THREAD

**Issue by llemeurfr** [readium/readium-lcp-client#43]
The call to LCPService openLicense always returns error code 8 (expired), whatever the cause of the error (expired, returned, revoked...)
-- Olivier Körner.

Therefore, the app cannot adapt the message to the status, which is bad.

**Comment by danielweck** [comment]
Culprit?
https://github.com/readium/readium-lcp-client/blob/273dc2b3dea9cfe059698928415ed8aabf95c35f/src/lcp-client-lib/RightsLcpNode.cpp#L76-L79

**Comment by danielweck** [comment]
Problem: the License JSON (and corresponding in-memory data model) does not store information about the reason *why* the LCP license is "expired" (i.e. `User.Rights.End` <= current date/time). The only way to know is to fetch the LSD JSON and read its status property. This is an asynchronous request totally disconnected from the core LCP processing. This additional information is only obtainable if: there is an LSD link in the LCP license, there is a network connection on the device, and the HTTP LSD request is successful.

**Comment by llemeurfr** [comment]
You're right, the license document does not contain the status of the license, only the status document does. In case of cancellation or revocation, the app will find in the license its expiration datetime  with a value before now, therefore "expired". 
Note that in production mode, LCP license MUST has an LSD link. When a user opens an ebook, the app checks the LSD link, fetches the Status Document, get the status, checks the date of update of the license, if updated fetches the license.  
 
So, to avoid a misleading information, the user message should al least read something like "This license is no longer active". And a user friendly app will store the status when the Status Document is accessed, so that it can display the proper status message, plus give access to the history of events if available.

**Comment by olivierkorner** [comment]
Note that an app doesn't have access to the status. There isn't any public property or method in LCPStatusDocumentProcessing giving access to the status, only isActive or wasCancelled.

**Comment by danielweck** [comment]
Agreed.
However, because the LSD JSON fetcher + LSD interaction handler is an asynchronous process in both Readium1 and 2, we need to add a data payload to some sort of a "ready" event that a consumer app can query live or even after-the-fact (there would need to be an implementation of that API for each target platform, using platform-native language, i.e. Readium1-Android-Java, Readium1-iOS-ObjectiveC, Readium2-ElectronNode-TypScript, etc.).
For instance, LSD information will only be available after the first time an EPUB gets opened, and the library view (i.e. EPUB collection with ebook covers, bibliographic metadata, purchase status, etc.) cannot be properly populated until after the full LCP "protocol" is complete.
An alternative option is for real-world apps to query the LCP data model for the currently-opened EPUB, get the LSD link (if any), fetch the LSD JSON, and parse the `status` field. But ideally this code should be provided by the off-the-shelve `readium-lcp-client` lib.

**Comment by olivierkorner** [comment]
Isn't it the role of the `StatusDocumentProcessingListener` protocol (on iOS)?
I understood that when the `onStatusDocumentProcessingComplete` method is called, the status has been updated.
Actually, I think the `LCPStatusDocumentProcessing` class is too opaque. It doesn't expose the links either, so I can't get the HTML renew or return links if I want to handle them by opening an webview.
As the library is provided as-is and I can't override or modify it, it would be nice to be able to get more of the status document information without having to re-implement 'get the LSD, fetch and parse'.

**Comment by danielweck** [comment]
Yes, `onStatusDocumentProcessingComplete` is what I am talking about. It is a common part of the LSD API in Android, iOS, etc. platform-native implementations. Problem: EPUBs have to be opened first, including a completed asynchronous round of LSD processing (with potential LCP license update = download + injection). So, a useful addition would be to expose a simple service that would "fetch the LSD from the link advertised in LCP". Either way, we need to implement this new functionality to give apps more flexibility for displaying LCP-LSD metadata, and to interact with LSD in an out-of-line fashion.

**Comment by danielweck** [comment]
Android internal class data which we could open publicly (available asynchronously via the `onStatusDocumentProcessingComplete` callback):
https://github.com/readium/readium-lcp-client/blob/develop/platform/android/lib/src/clientlib/java/org/readium/sdk/lcp/StatusDocumentProcessing.java#L266
iOS:
https://github.com/readium/readium-lcp-client/blob/develop/platform/apple/src/LCPStatusDocumentProcessing.mm#L116

**Comment by danielweck** [comment]
Android callback:
https://github.com/readium/SDKLauncher-Android/blob/feature/lcp/SDKLauncher-Android/app/src/main/java/org/readium/sdk/android/launcher/ContainerList.java#L519

iOS callback:
https://github.com/readium/SDKLauncher-iOS/blob/feature/lcp/Classes/ContainerController.mm#L161

**Comment by danielweck** [comment]
@olivierkorner you say "As the library is provided as-is and I can't override or modify it", but in your app integration you can actually modify the LSD "processor", right? I mean, it is not part of the pre-built C++ lib, it is open-source code written in platform code. If so, can you please run some tests with publicly-accessible getters for the LSD status, links, etc., so that you can provide implementation feedback, and we can fine-tune the API to benefit everybody based on real-world integration requirements. Meanwhile I will see what we can do to fetch LSD info separately from the actual EPUB opening phase.

**Comment by olivierkorner** [comment]
Mmhh you're right. 
Actually, the LCPStatusDocumentProcessing is built into the C++ lib but I don't have to use it! I can use my own. Thank you for your insight!

**Comment by danielweck** [comment]
Oh that's right, for iOS the `LCPStatusDocumentProcessing.h / mm` source is compiled into the pre-built `readium-lcp-client` lib (which includes ObjectiveC as well as C/C++ code).
https://github.com/readium/readium-lcp-client/blob/develop/platform/apple/LCP%20Client.xcodeproj/project.pbxproj

**Comment by danielweck** [comment]
It's a different development story on Android, as the Java code is separate from the C/C++ JNI code:
https://github.com/readium/readium-lcp-client/tree/develop/platform/android/lib/src/clientlib

**Comment by danielweck** [comment]
@olivierkorner `LCPStatusDocumentProcessing` is open-source on GitHub so practically-speaking you should be able to fork the `readium-lcp-client` repository and contribute your changes via a Pull Request back to:
https://github.com/readium/readium-lcp-client/blob/develop/platform/apple/src/LCPStatusDocumentProcessing.mm
https://github.com/readium/readium-lcp-client/blob/develop/platform/apple/src/LCPStatusDocumentProcessing.h
Let us know what additions you make to the public API, and also if you implement an alternative technique to fetch the LSD (from the link advertised in the LCP license) *before* the EPUB is fully loaded (i.e. a simple LSD getter that does not attempt to update+inject the LCP license when there is one available). Other thoughts welcome too :)
